src/coreclr/System.Private.CoreLib/src/System/Diagnostics/StackTrace.CoreCLR.cs

```diff
@@ -66,30 +71,55 @@ private void CaptureStackTrace(int skipFrames, bool fNeedFileInfo, Exception? e)
             StackF.InitializeSourceInfo(fNeedFileInfo, e);
 
             _numOfFrames = StackF.GetNumberOfFrames();
+            _stackFrameHelper = StackF;
 
             if (_methodsToSkip > _numOfFrames)
                 _methodsToSkip = _numOfFrames;
 
-            if (_numOfFrames != 0)
+            _fNeedFileInfo = fNeedFileInfo;
+
+            // CalculateFramesToSkip skips all frames in the System.Diagnostics namespace,
+            // but this is not desired if building a stack trace from an exception.
+            if (e == null)
+                _methodsToSkip += CalculateFramesToSkip(_stackFrameHelper, _numOfFrames);
+
+            _numOfFrames -= _methodsToSkip;
+            if (_numOfFrames < 0)
             {
-                _stackFrames = new StackFrame[_numOfFrames];
+                _numOfFrames = 0;
+            }
 
-                for (int i = 0; i < _numOfFrames; i++)
-                {
-                    _stackFrames[i] = new StackFrame(StackF, i, fNeedFileInfo);
-                }
+            // Grab MethodBases for all frames eagerly to keep them alive, in case one of them
```

Comment:
I am looking at this again. Should `object? dynamicMethods` field on StackFrameHelper be enough to keep things alive?

Reply:
It looks like it, actually. It seems native stores the loader allocator object for collectable MTs there. I'll remove this here and update some comments to make that more obvious.